src/ray/gcs/gcs_server/gcs_health_check_manager.h

```diff
@@ -0,0 +1,151 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <grpcpp/grpcpp.h>
+
+#include "absl/container/flat_hash_map.h"
+#include "ray/common/asio/instrumented_io_context.h"
+#include "ray/common/id.h"
+#include "ray/common/ray_config.h"
+#include "src/proto/grpc/health/v1/health.grpc.pb.h"
+
+class GcsHealthCheckManagerTest;
+
+namespace ray {
+namespace gcs {
+
+/// GcsHealthCheckManager is used to track the healthiness of the nodes in the ray
+/// cluster. The health check is done in pull based way, which means this module will send
+/// health check to the raylets to see whether the raylet is healthy or not. If the raylet
+/// is not healthy for certain times, the module will think the raylet dead.
+class GcsHealthCheckManager {
+ public:
+  /// Constructor of GcsHealthCheckManager.
+  ///
+  /// \param io_service The thread where all operations in this class should run.
+  /// \param on_node_death_callback The callback function when some node is marked as
+  /// failure. \param initial_delay_ms The delay for the first health check. \param
+  /// period_ms The interval between two health checks for the same node. \param
+  /// failure_threshold The threshold before a node will be marked as dead due to health
```

Comment:
failure_threshold -> num_consecutive_failures_threshold?

Reply:
I borrow the terminology from k8s (https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes)

Since the whole protocol also has refer a lot from that, I think maybe we should keep it? 

(no strong opinion).